client/core/sshclient.cpp

```diff
@@ -222,99 +222,89 @@ namespace libssh {
+            int localFileSize = QFileInfo(localPath).size();
 
-            if (file == nullptr) {
-                return closeSftpSession();
+            int rc = ssh_scp_push_file(m_scpSession, remotePath.toStdString().c_str(), localFileSize, accessType);
+            if (rc != SSH_OK) {
+                return closeScpSession();
             }
 
-            int localFileSize = QFileInfo(localPath).size();
-            int chunksCount = localFileSize / (bufferSize);
-
+            int chunksCount = localFileSize / bufferSize;
             QFile fin(localPath);
 
             if (fin.open(QIODevice::ReadOnly)) {
                 for (int currentChunkId = 0; currentChunkId < chunksCount; currentChunkId++) {
                     QByteArray chunk = fin.read(bufferSize);
-                    if (chunk.size() != bufferSize) return ErrorCode::SshSftpEofError;
-
-                    int bytesWritten = sftp_write(file, chunk.data(), chunk.size());
+                    if (chunk.size() != bufferSize) {
+                        fin.close();
+                        return ErrorCode::InternalError;
+                    }
 
-                    if (bytesWritten != chunk.size()) {
+                    rc = ssh_scp_write(m_scpSession, chunk.data(), chunk.size());
+                    if (rc != SSH_OK) {
                         fin.close();
-                        sftp_close(file);
-                        return closeSftpSession();
+                        return closeScpSession();
                     }
                 }
 
                 int lastChunkSize = localFileSize % bufferSize;
 
                 if (lastChunkSize != 0) {
                     QByteArray lastChunk = fin.read(lastChunkSize);
-                    if (lastChunk.size() != lastChunkSize) return ErrorCode::SshSftpEofError;
-
-                    int bytesWritten = sftp_write(file, lastChunk.data(), lastChunkSize);
+                    if (lastChunk.size() != lastChunkSize) {
+                        fin.close();
+                        return ErrorCode::InternalError;
+                    }
 
-                    if (bytesWritten != lastChunkSize) {
+                    rc = ssh_scp_write(m_scpSession, lastChunk.data(), lastChunk.size());
+                    if (rc != SSH_OK) {
                         fin.close();
-                        sftp_close(file);
-                        return closeSftpSession();
+                        return closeScpSession();
                     }
                 }
             } else {
-                sftp_close(file);
-                return closeSftpSession();
+                return closeScpSession();
             }
 
             fin.close();
 
-            int result = sftp_close(file);
-            if (result != SSH_OK) {
-                return closeSftpSession();
-            }
-
-            return closeSftpSession();
+            return closeScpSession();
```

Comment:
Looks like you need to review and simplify whole scpFileCopy logic.
For example, closeScpSession() missing in some function exit conditions, but QFile fin not necessary to close at all.